emspy/connection.py

```diff
@@ -1,184 +1,296 @@
-
-		# uri = uri.encode('utf-8')
-		req = urllib.request.Request(uri, data=data, headers=headers)
-		try:
-			resp = self.__send_request(req)
-			statcode = resp.getcode()
-			if statcode!=200:
-				print("Http status code: %d" % statcode)
-				verbose = True
-		except:
-			(eType, eValue, eTrace) = sys.exc_info()
-			if eType is ssl.CertificateError:
-				print("A certificate verification error occured for the request to '%s'. Certificate verification is required by default, but can be disabled by using the ignore_ssl_errors argument for the Connection constructor." % uri )
-				raise
-			if eType is urllib.error.HTTPError:
-				message_bytes = eValue.read()
-				message_str = message_bytes.decode('utf-8')
-				eValue.msg = eValue.msg + '\n Details: ' + message_str
-				raise
-
-			print("Trying to reconnect the EMS API.")
-			self.reconnect()
-			print("Done.")
-			resp = self.__send_request(req)
-		resp_h   = resp.getheaders()
-
-		# If the response is compressed, decompress it.
-		if resp.info().get('Content-Encoding') == 'gzip':
-			buffer = io.BytesIO(resp.read())
-			file = gzip.GzipFile(fileobj=buffer)
-			content = json.loads(file.read())
-		else:
-			content = json.loads(resp.read())
-			
-		if verbose:
-			print("URL: %s" % resp.geturl())
-			pp.pprint(resp_h)
-			pp.pprint(content)
-
-		return resp_h, content
-
-
-	def __send_request(self, req):
-		"""Sends the request and returns the response, optionally ignoring ssl errors."""
-
-		# Normally you do NOT want to ignore SSL errors, but this is
-		# sometimes necessary on beta API endpoints without a proper cert.
-		# TODO: Find a real fix for GE Mac OS machines that are having trouble verifying.  
-		request = None
-		if self.__ignore_ssl_errors:
-			try:
-				request = urllib.request.urlopen(req, context=None)
-			except urllib.error.URLError:
-				request = urllib.request.urlopen(req, context=ssl._create_unverified_context())
-		else:
-			request = urllib.request.urlopen(req, context=None)
-		return request
-
-
-def print_resp(resp):
-	
-	for r in resp:
-		pp.pprint(r)
+    """
+    Object for connection to EMS API
+    """
+    def __init__(self, user=None, pwd=None, proxies=None, verbose=False, ignore_ssl_errors=False,
+                 server="prod", server_url=None, max_trials=3):
+        """
+        Connection initialization
+
+        Parameters
+        ----------
+        user: str
+            EMS username
+        pwd: str
+            EMS password
+        proxies: dict
+            proxies dictionary {'http': '', 'https': ''}
+        verbose: bool
+            verbose output (default False)
+        ignore_ssl_errors: bool
+            ignore SSL certificate errors (default False)
+        server: str
+            server to connect to from emspy.common.uri_root
+        server_url: str
+            hostname for server to connect to if different from any of the common values
+        max_trials: int
+            maximum number of reconnection attempts
```

Comment:
Do we want to specify which parameters are optional?

Reply:
I think the class signature should make it evident which parameters are optional, but I can add the optional tags to all functions (would take me a while)